Implementation of labeling engine rules

nyalldawson opened the pull request:
See https://github.com/qgis/QGIS-Enhancement-Proposals/issues/299

Implements the API framework for setting advanced labeling engine rules for a project, and implements 4 initial rule types:

- QgsLabelingEngineRuleMinimumDistanceLabelToFeature: prevents labels   being placed too *close* to features from a different layer
- QgsLabelingEngineRuleMaximumDistanceLabelToFeature: prevents labels   being placed too *far* from features from a different layer
- QgsLabelingEngineRuleMinimumDistanceLabelToLabel: prevents labels   being placed too close to labels from a different layer
- QgsLabelingEngineRuleAvoidLabelOverlapWithFeature: prevents labels   being placed overlapping features from a different layer

(note that the first 3 rules require a build based on GEOS >= 3.10, they are not available for older GEOS builds)

Also implements a registry for storing available rule classes, and serialization of rules and configuration in QGIS projects

Now with GUI:


https://github.com/user-attachments/assets/a8854ce4-4a55-4c2f-abf9-8cc68517ceae



Sponsored by Rubicon Concierge Real Estate

nyalldawson committed:
[api] Implementation of labeling engine rules

See https://github.com/qgis/QGIS-Enhancement-Proposals/issues/299

Implements the API framework for setting advanced labeling engine
rules for a project, and implements 4 initial rule types:

- QgsLabelingEngineRuleMinimumDistanceLabelToFeature: prevents labels
  being placed too *close* to features from a different layer
- QgsLabelingEngineRuleMaximumDistanceLabelToFeature: prevents labels
  being placed too *far* from features from a different layer
- QgsLabelingEngineRuleMinimumDistanceLabelToLabel: prevents labels
  being placed too close to labels from a different layer
- QgsLabelingEngineRuleAvoidLabelOverlapWithFeature: prevents labels
  being placed overlapping features from a different layer

(note that the first 3 rules require a build based on GEOS >= 3.10,
they are not available for older GEOS builds)

Also implements a registry for storing available rule classes,
and serialization of rules and configuration in QGIS projects

nyalldawson committed:
pal cleanups - better variable names, comments

nyalldawson committed:
Respect engine rules when solving pal problem

Adds the necessary hooks into the pal classes to give engine rules
a chance to implement their logic

nyalldawson committed:
Add tests for rules

nyalldawson committed:
Fix seealso

nyalldawson committed:
Fix clang-tidy warnings

nyalldawson committed:
Rules can have user-configurable names

nyalldawson committed:
Add tests

nyalldawson committed:
Add __repr__

nyalldawson committed:
Add gui for configuring project advanced labeling rules

nyalldawson committed:
Add icon for labeling rules

nyalldawson committed:
Don't lose other label settings when configuring engine settings

nyalldawson committed:
Correctly update rules when rule removed

nyalldawson commented:
Now with GUI, this is ready for review...

nyalldawson committed:
Fix docs

DelazJ commented:
Nice @nyalldawson! What about showing involved layers next to the rule in the "Labeling rules" list? It might help to quickly identify the rule you want to update...

nyalldawson commented:
@DelazJ you can set your own name for rules, which is used in that list. I was thinking that could be better for rule identification.

Maybe we could show a summary including involved layers in a tooltip for the list item?

DelazJ commented:
Ah, great! I only watched the screencast which didn't showcase that.
> Maybe we could show a summary including involved layers in a tooltip for the list item?

+1 from me. When you are out of inspiration for rules naming, this could be handful 😄

nyalldawson committed:
Allow labeling rules to be de-activated

nyalldawson committed:
Rule name can be edited via list

nyalldawson committed:
Generate user-friendly tooltips

nyalldawson commented:
@DelazJ 
here you go:

https://github.com/user-attachments/assets/a22f3f06-4110-47fe-bdf8-5bd4c32a22c1

nirvn commented on src/core/labeling/rules/qgslabelingenginerule.h:
```diff
@@ -0,0 +1,299 @@
+     *
+     * The returned string may contain HTML formatting.
+     */
+    virtual QString description() const;
```
An icon representation could be nice too (not blocking this PR for that, just food for thoughts).

nirvn commented on src/core/labeling/rules/qgslabelingenginerule.h:
```diff
@@ -0,0 +1,299 @@
+     *
+     * The default implementation returns FALSE.
+     */
+    virtual bool candidateIsIllegal( const pal::LabelPosition *candidate, QgsLabelingEngineContext &context ) const SIP_SKIP;
```
I'm not sure why but I don't like the word 'illegal' here. What about candidateIsForbidden, candiateIsProhibited, candidateIsDisallowed, candidateIsBanned ? We're breaching rules not laws :wink:

nyalldawson replied:
Hmm... we have `QgsAbstractDatabaseProviderConnection::illegalFieldNames`, but nothing in current api with forbidden/prohibited/disallowed/banned. Is one other method enough to set the precendent?

nirvn replied:
@nyalldawson , ideally we'd change illegalFieldNames to prohibited/forbiddenFileNames but that API is exposed to bindings so it's frozen right? 

Seems like settling for what preexisting keyword is a good argument enough then.

nirvn commented on src/core/labeling/rules/qgslabelingenginerule_impl.cpp:
```diff
@@ -0,0 +1,632 @@
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgslabelingenginerule_impl.h"
+#include "qgsunittypes.h"
+#include "qgssymbollayerutils.h"
+#include "labelposition.h"
+#include "feature.h"
+#include "qgsvectorlayerfeatureiterator.h"
+#include "qgsthreadingutils.h"
+#include "qgsspatialindex.h"
+#include "qgsgeos.h"
```
```suggestion
#include "qgslabelingenginerule_impl.h"
#include "qgsunittypes.h"
#include "qgssymbollayerutils.h"
#include "qgsvectorlayerfeatureiterator.h"
#include "qgsthreadingutils.h"
#include "qgsspatialindex.h"
#include "qgsgeos.h"
#include "labelposition.h"
#include "feature.h"
```

nirvn commented on src/core/labeling/rules/qgslabelingenginerule_impl.cpp:
```diff
@@ -0,0 +1,632 @@
+
+QString QgsLabelingEngineRuleMaximumDistanceLabelToFeature::displayType() const
+{
+  return QObject::tr( "Pull Labels toward Features" );
```
I'm pretty sure one must capitalize toward(s) in a title case sentence.

```suggestion
  return QObject::tr( "Pull Labels Toward Features" );
```

nirvn commented on src/core/labeling/rules/qgslabelingengineruleregistry.cpp:
```diff
@@ -0,0 +1,81 @@
+
+QgsLabelingEngineRuleRegistry::QgsLabelingEngineRuleRegistry()
+{
+#if GEOS_VERSION_MAJOR>3 || ( GEOS_VERSION_MAJOR == 3 && GEOS_VERSION_MINOR>=10 )
+  addRule( new QgsLabelingEngineRuleMinimumDistanceLabelToFeature() );
+  addRule( new QgsLabelingEngineRuleMaximumDistanceLabelToFeature() );
+  addRule( new QgsLabelingEngineRuleMinimumDistanceLabelToLabel() );
+#endif
```
If we're conditionally adding rules to the registry based on GEOS version, that'll mean a project shared across multiple users with different GEOS versions (or projects processed on a server) will result in rule loss. Is that something you've thought about and are we OK with that? 

An alternative would be to add a QLabel note in the rule widget UI saying 'this rule will be ignored on your system' or something.

nyalldawson replied:
Ok, I've gone with:

- grey out the rule in the rules list
- show an explanatory tooltip
- don't allow users to create new versions of that rule

nirvn commented on src/core/pal/pal.cpp:
```diff
@@ -396,17 +428,18 @@ std::unique_ptr<Problem> Pal::extractProblem( const QgsRectangle &extent, const
       conflictProfile = std::make_unique< QgsScopedRuntimeProfile >( QObject::tr( "Calculating conflicts" ), QStringLiteral( "rendering" ) );
     }
 
-    int idlp = 0;
```
Thanks for cleaning up these obscure variable names.

nirvn commented on src/core/pal/pal.h:
```diff
@@ -133,6 +134,17 @@ namespace pal
+       * - collects obstacles
+       * - calculates candidate costs
+       * - calculates overlaps/conflicts
+       * - eliminates hard conflicts (forbidden placement)
```
Forbidden! Nice nice, I'd change illegal to forbidden :)

nyalldawson committed:
Sipify

nyalldawson committed:
Address review

nyalldawson committed:
If a rule is not available on the system (eg geos too old):

- grey out the rule in the rules list
- show an explanatory tooltip
- don't allow users to create copies of that rule
